Add Reserved Keywords document (#45)
partiql/partiql-docs: docs/reserved-keywords.adoc

= PartiQL Reserved Keywords

> Reserved keywords are part of PartiQL Language that are used by PartiQL to parse and understand PartiQL statements.

Reviewer: Sentence phrasing is slightly awkward

```suggestion
The PartiQL language uses reserved keywords to parse and understand PartiQL statements.
```

PartiQL Language Grammar Reserved Keywords cannot be used in PartiQL statements as-is.
For example, the following query results in a _Syntax Error_ because `FROM` is a Reserved Keyword in PartiQL:

.Example 1. Invalid PartiQL Query with a Reserved Keyword
[source, sql]
----
SELECT t.FROM FROM <<{'ID': 1, 'FROM': 'NA'}, {'ID': 2, 'FROM': 'SA'}>> AS t;
----

In order to use Reserved Keywords in your PartiQL statements, use Quoted Identifiers as Example 2 shows:

.Example 2. Using Quoted Identifier as a Reserve Keyword in a PartiQL Statement.
[source, sql]
----

> SELECT t."FROM" FROM <<{'ID': 1, 'FROM': 'NA'}, {'ID': 2, 'FROM': 'SA'}>> AS t;

Reviewer: Probably want to also call out that the reserved keyword variable reference won't be case sensitive since you use double quotes.

Author: Makes sense; I'll point that out.

Also as they are going to be case-sensitive after putting between double-quotes; I'll clarify that too:

```
PartiQL> SELECT DISTINCT t."FROM" FROM <<{'ID': 1, 'from': 'NA'}, {'ID': 2, 'from': 'SA'}>> AS t;
==='
<<
  {}
>>
---
OK!

PartiQL> SELECT DISTINCT t."from" FROM <<{'ID': 1, 'FROM': 'NA'}, {'ID': 2, 'FROM': 'SA'}>> AS t;
==='
<<
  {}
>>
---
OK!

PartiQL> SELECT DISTINCT t."from" FROM <<{'ID': 1, 'FROM': 'NA'}, {'ID': 2, 'from': 'SA'}>> AS t;
==='
<<
  {}
>>
---
OK!
```

----

== Reserved Keywords
[NOTE]
====

> Reserved Keywords are yet to be https://github.com/partiql/partiql-spec/issues/2[formally defined] in PartiQL Specification, therefore there may be some differences in different PartiQL implementations.

Reviewer: I think it could be helpful to make a version of this document for each PartiQL version and implementation listing the reserved keywords (at least the versions with differences).

This can be useful as PartiQL versions change and develop over time. I know we've had similar questions related to our builtin functions (e.g. does partiql-lang-kotlin-x.y.z support this builtin function?).

For reference Microsoft's SQL server has different pages for their different versions:
- SQL-server 2022 -- https://learn.microsoft.com/en-us/sql/t-sql/language-elements/reserved-keywords-transact-sql?view=sql-server-ver16
- SQL-server 2019 -- https://learn.microsoft.com/en-us/sql/t-sql/language-elements/reserved-keywords-transact-sql?view=sql-server-ver15
- ...

Author: I think putting two additional dimensions (i.e., implementation and version) brings too much overhead unless we have a clear spec. definition for this. I see this document's objective to give a general list of possible keywords and a way to escape them while mentioning that the actual keywords can be different.

Author: Perhaps a good compromise is adding PartiQL Spec. Version?

Reviewer: I agree that it could add overhead. I still think we should document somewhere though each implementation + implementation version's reserved keywords since they may change and add more keywords over time. Perhaps for each implementation we can have a doc listing the reserved keywords. As we perform new tagged releases on GitHub, we'll then have a link to the reserved keywords of that specific PartiQL version.

E.g. if we had a document called reserved-keywords.adoc in partiql-lang-kotlin, we could reference it like:
- https://github.com/partiql/partiql-lang-kotlin/tree/vx.y.z-alpha/docs/reserved-keywords.adoc
And for Rust do something similar:
- https://github.com/partiql/partiql-lang-rust/tree/vx.y.z-alpha/docs/reserved-keywords.adoc

Author: I generally agree, but I think we need to automate this using the Grammar in the code.

Reviewer: Perhaps we could add an issue for each implementation to
- make a reserved keywords doc listing all the reserved keywords
- have some way to automate this doc?
I think it can still be helpful to have the reserved keywords of each implementation/version visible

Author: Just reflecting the offline response:

I think the following note should be sufficient:
```
Reserved keywords are yet to be https://github.com/partiql/partiql-spec/issues/2[formally defined] in PartiQL Specification, therefore there may be some differences in different PartiQL implementations.
```

I’m hesitant to go to a lower level b/c that increases the likelihood of exposing stale information if for example, Rust grammar changes.

I still think we need a story around versioned docs for implementations though and I’m going to open an issue for that.

Will update this thread with the issue.

Author: Here is the issue for tracking the version-based documentation:
https://github.com/partiql/partiql-docs/issues/46

====

[cols="1,1,1,1",grid=none, frame=none]
|===

> |ABSOLUTE

Reviewer: Out of curiosity, where was this initial list taken from?

Author: It's from here: https://partiql.org/syntax/antlr.html
